python/cuml/cuml/manifold/t_sne.pyx

```diff
@@ -141,6 +132,128 @@ _SUPPORTED_METRICS = {
     "correlation": DistanceType.CorrelationExpanded
 }
 
+_SUPPORTED_METHODS = {
+    "barnes_hut": TSNE_ALGORITHM.BARNES_HUT,
+    "exact": TSNE_ALGORITHM.EXACT,
+    "fft": TSNE_ALGORITHM.FFT,
+}
+
+_SUPPORTED_INITS = {
+    "random": TSNE_INIT.RANDOM,
+    "pca": TSNE_INIT.PCA,
+}
+
+
+def _check_numeric(estimator, name, gt=None, ge=None, lt=None, le=None):
+    """Check that a numeric parameter `name` is within valid bounds"""
+    value = getattr(estimator, name)
+    cls_name = type(estimator).__name__
+    if gt is not None and value <= gt:
+        raise ValueError(f"{cls_name} requires `{name} > {gt}`, got {value}")
+    if ge is not None and value < ge:
+        raise ValueError(f"{cls_name} requires `{name} >= {ge}`, got {value}")
+    if lt is not None and value >= lt:
+        raise ValueError(f"{cls_name} requires `{name} < {lt}`, got {value}")
+    if le is not None and value > le:
+        raise ValueError(f"{cls_name} requires `{name} <= {le}`, got {value}")
+    return value
+
+
+def _check_mapping(estimator, name, mapping):
+    """Check that a parameter `name` contained within a valid mapping"""
+    value = getattr(estimator, name)
+    cls_name = type(estimator).__name__
+    try:
+        return mapping[value]
+    except KeyError:
+        raise ValueError(
+            f"{cls_name} expects `{name}` to be one of {sorted(mapping)}, got {value}"
+        ) from None
+
+
+cdef _init_params(self, int n_samples, TSNEParams &params):
+    """Validate TSNE parameters and initialize a TSNEParams instance."""
+    if (n_components := self.n_components) != 2:
+        raise ValueError(
+            f"Currently TSNE only supports n_components = 2, got {self.n_components}"
+        )
+    perplexity = _check_numeric(self, "perplexity", gt=0)
```

Comment:
We previously failed when `perplexity < 0`; now we fail for `perplexity <= 0`. I assume that's intentional. Did you double-check and adjust any of the other parameter boundaries? For example, `learning_rate` also changed from `>= 0` to `> 0` (with the latter being correct AFAIK).

Reply:
I've double-checked the entries and they look correct to me. I am wondering why `early_exaggeration==0` is valid given that we divide by it, but I didn't encounter any issues while testing. 🤷🏻 

https://github.com/rapidsai/cuml/blob/6166b724af13918724da5c08bcf55daa3aaf7314/cpp/src/tsne/barnes_hut_tsne.cuh#L186